File: LeetCode June Challenge/LongestStringChain.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
private:
    static bool sortByLength(string &word1, string &word2){
        return word1.size() < word2.size();
    }
public:
    int longestStrChain(vector<string>& words) {
        sort(words.begin(), words.end(), sortByLength);
        unordered_map<string, int> dp; int ans = 0;
        for(auto word : words){
            for(int i = 0; i < word.size(); ++i){
                string predecessor = word.substr(0, i) + word.substr(i + 1);
                dp[word] = max(dp[word], dp[predecessor] + 1);
            }
            ans = max(ans, dp[word]);
        }
        return ans;
    }
};
int main(){
    vector<string> words = {"a","b","ba","bca","bda","bdca"};
    Solution s;
    cout<<s.longestStrChain(words)<<endl;
    return 0;
}

File: Year-2022/LeetCode June Challenge/LongestStringChain.js
var longestStrChain = function(words) {
    words.sort((a, b) => a.length - b.length);
    let dp = {}; let ans = 0;
    for (let word of words){
        let len = 0;
        for(let i = 0; i < word.length; ++i){
            let predecessor = word.slice(0, i) + word.slice(i + 1);
            len = Math.max(len, (dp[predecessor] || 0) + 1);
        }
        dp[word] = Math.max(ans, len);
    }
    return Math.max(...Object.values(dp));
};
let words = ["a","b","ba","bca","bda","bdca"];
console.log(longestStrChain(words));